### PR TITLE
feat(environmentConfig): allow setting sourceFieldPath label selector optional

### LINKED ISSUE #3723: Allow ignoring environment config if it does not exists
### What problem are you facing?

I don't know if I am using environment config the bad way but I can't figure how to ignore selection of an `EnvironmentConfig` if it does not exists.

Let's say I have the following `EnvironmentConfig` resources:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: EnvironmentConfig
metadata:
  labels:
    provider-name: default-provider
  name: defaut-provider-config
data:
  id: 123456
 ```
 
```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: EnvironmentConfig
metadata:
  labels:
    provider-name: other-provider
  name: other-provider-config
data:
  id: 456789
 ```

Then I have a XRD with an optional field `providerConfigRef.name` in spec

In my composition I want to load the `default-provider-config EnvironmentConfig` if `providerConfigRef.name` is not referenced in my XR, and the correct provider `EnvironmentConfig` if it is referenced (let's say `other-provider-config EnvironmentConfig` if `providerConfigRef.name : other-provider`). 

So I tried to add the following environment in the associated composition:

```yaml
...
environment:
  environmentConfigs:
    - type: Selector
      selector:
        matchLabels:
        - key: provider-name
          type: Value
          value: default-provider
    - type: Selector
      selector:
        matchLabels:
        - key: provider-name
          type: FromCompositeFieldPath
          valueFromFieldPath: spec.providerConfigRef.name
```

The problem is if `spec.providerConfigRef.name` does not exists, crossplane will fail reconciliation with the following message: `cannot select environment: failed to build reference at index 1: failed to resolve value for label at index 0: spec.providerConfigRef: no such field` 

### How could Crossplane help solve your problem?

Maybe there is a workaround to this issue, but if not it would be neat to have a way to ignore missing environment config, the same way as patches are ignored if field path does not exists

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Fixes #3723 .

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR, if necessary.
- [x] Opened a PR updating the [docs], if necessary.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

**Review by MisterMX — APPROVED** [review]

File: internal/controller/apiextensions/composite/environment/selector.go [hunk 1]
```diff
@@ -132,7 +135,7 @@ func (s *APIEnvironmentSelector) lookUpConfigs(ctx context.Context, cr resource.
 func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromSelector(cl *v1alpha1.EnvironmentConfigList, selector *v1.EnvironmentSourceSelector) ([]corev1.ObjectReference, error) {
 	ec := make([]v1alpha1.EnvironmentConfig, 0)
 
-	if len(cl.Items) == 0 {
+	if cl == nil || len(cl.Items) == 0 {
```
Comment (pedjak) [inline internal/controller/apiextensions/composite/environment/selector.go:138]: should be a regular case to pass `nil` for `cl` in this function? `lookUpConfig` never returns `nil`, `nil` so perhaps we should return error instead? Otherwise, could we add selector unit test for this change?

File: internal/controller/apiextensions/composite/environment/selector_test.go [hunk 2]
```diff
@@ -510,6 +511,41 @@ func TestSelect(t *testing.T) {
 				err: errors.Wrapf(errors.Wrapf(errors.New("wrong: no such field"), errFmtResolveLabelValue, 0), errFmtReferenceEnvironmentConfig, 0),
 			},
 		},
+		"NoErrorOnInvalidOptionalLabelValueFieldPath": {
+			reason: "It should not return an error if the path to a label value is invalid, but was set as optional.",
+			args: args{
+				kube: &test.MockClient{
+					MockList: test.NewMockListFn(nil),
```
[the remark this answers is not in the document]
Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector_test.go:518]: It's enough to not define it and it will panic.

File: internal/controller/apiextensions/composite/environment/selector.go (hunk 1 above)
Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector.go:138]: It should never happen but I feel it's ok to handle it as if it had 0 items.

File: internal/controller/apiextensions/composite/environment/selector_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector_test.go]: updated test accordingly

[the remark this answers is not in the document]
Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector_test.go]: However I would argue that would not be a regression, I don't care if under the hood the function calls a list or not before returning, I only care this one is accepted, while it's erroring out if set to required. But that's mostly a matter of taste and mocking does push toward such an approach.

[the remark this answers is not in the document]
Comment (pedjak) [reply internal/controller/apiextensions/composite/environment/selector_test.go]: IMHO, checking how the component under tests interacts with dependencies is important (on unit-test level). If a path does not exist and the label cannot be constructed, then interacting with cluster should never happen.

File: internal/controller/apiextensions/composite/environment/selector.go (hunk 1 above)
Comment (pedjak) [reply internal/controller/apiextensions/composite/environment/selector.go:138]: If it should never happen, then passing `nil` here might smell like a bug, and failing early would uncover it. 

https://en.wikipedia.org/wiki/Design_by_contract

Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector.go:138]: Then let me rephrase: "it can happen and should be handled as if it was an empty list"

Comment (pedjak) [reply internal/controller/apiextensions/composite/environment/selector.go:138]: ok, can you please add selector unit test for that situation?

**Review by pedjak — APPROVED** [review]

File: apis/apiextensions/v1/composition_environment.go [hunk 3]
```diff
@@ -239,10 +239,23 @@ type EnvironmentSourceSelectorLabelMatcher struct {
 	// ValueFromFieldPath specifies the field path to look for the label value.
 	ValueFromFieldPath *string `json:"valueFromFieldPath,omitempty"`
 
+	// FromFieldPathPolicy specifies how to patch from a field path. The default is
+	// 'Required', which means the patch should fail if the path specified via valueFromFieldPath does not exist.
+	// Use 'Ignore' if instead you want it to result in a no-op.
+	// +kubebuilder:validation:Enum=Optional;Required
```
Comment (turkenh) [inline apis/apiextensions/v1/composition_environment.go:245]:
```suggestion
	// +kubebuilder:validation:Enum=Optional;Required
	// +kubebuilder:default=Required
```

File: apis/apiextensions/v1/composition_environment.go [hunk 4]
```diff
@@ -239,10 +239,23 @@ type EnvironmentSourceSelectorLabelMatcher struct {
 	// ValueFromFieldPath specifies the field path to look for the label value.
 	ValueFromFieldPath *string `json:"valueFromFieldPath,omitempty"`
 
+	// FromFieldPathPolicy specifies how to patch from a field path. The default is
+	// 'Required', which means the patch should fail if the path specified via valueFromFieldPath does not exist.
+	// Use 'Ignore' if instead you want it to result in a no-op.
```
Comment (turkenh) [inline apis/apiextensions/v1/composition_environment.go:244]: Is it `Optional vs Required` or `Ignore vs Required`?

File: apis/apiextensions/v1/composition_environment.go [hunk 5]
```diff
@@ -239,10 +239,23 @@ type EnvironmentSourceSelectorLabelMatcher struct {
...
+	// +kubebuilder:validation:Enum=Optional;Required
+	// +optional
+	FromFieldPathPolicy *FromFieldPathPolicy `json:"policy,omitempty"`
```
Comment (turkenh) [inline apis/apiextensions/v1/composition_environment.go:247]:
```suggestion
	FromFieldPathPolicy *FromFieldPathPolicy `json:"fromFieldPathPolicy,omitempty"`
```

Should we make it more obvious?

Otherwise, this looks like below in the API, right?

```yaml
          matchLabels:
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: spec.someValue
              policy: Optional
```

This looks like matching this label (i.e. `matchLabels[0]`) is `Optional`, which I believe is kind of true if `spec.someValue` is not there, we ignore this label. However, it could be confusing in the sense that, even that value is there, this label is optional, which doesn't make sense though. For example, things get interesting if this is used in conjunction with `value` instead:

```yaml
        matchLabels:
          - type: Value
            key: stage
            value: prod
            policy: Optional
```

File: cluster/crds/apiextensions.crossplane.io_compositionrevisions.yaml [hunk 6]
```diff
@@ -1603,6 +1614,17 @@ spec:
                                   key:
                                     description: Key of the label to match.
                                     type: string
+                                  policy:
+                                    description: FromFieldPathPolicy specifies how
```
Comment (turkenh) [inline cluster/crds/apiextensions.crossplane.io_compositionrevisions.yaml:1618]: `policy` vs `FromFieldPathPolicy` in the description of the API

File: internal/controller/apiextensions/composite/environment/selector.go [hunk 7]
```diff
@@ -114,15 +114,18 @@ func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromRef(ref *v1.Enviro
 }
 
 func (s *APIEnvironmentSelector) lookUpConfigs(ctx context.Context, cr resource.Composite, ml []v1.EnvironmentSourceSelectorLabelMatcher) (*v1alpha1.EnvironmentConfigList, error) {
+	res := &v1alpha1.EnvironmentConfigList{}
 	matchLabels := make(client.MatchingLabels, len(ml))
 	for i, m := range ml {
 		val, err := ResolveLabelValue(m, cr)
 		if err != nil {
+			if fieldpath.IsNotFound(err) && m.FromFieldPathIsOptional() {
+				return res, nil
```
Comment (turkenh) [inline internal/controller/apiextensions/composite/environment/selector.go:123]: Shouldn't we `continue` here instead of returning res which effectively has an empty environment config list?

Comment (phisco) [reply internal/controller/apiextensions/composite/environment/selector.go:123]: I had the same doubt, I decided to make it "no-op" the whole environment, as I thought that omitting a single value would have been more confusing. This was my thought:
- if a single label selector is specified and marked as optional, we should return no environment at all.
- if `n>1` selectors are specified, one of them is marked as optional, if not found I'd rather have it result in a "no-op" than selecting an environment selected only through the `n-1` remaining selectors.
But I agree it's debatable, what would you suggest?

Comment (turkenh) [reply internal/controller/apiextensions/composite/environment/selector.go:123]: > if n>1 selectors are specified, one of them is marked as optional, if not found I'd rather have it result in a "no-op" than selecting an environment selected only through the n-1 remaining selectors.

Since the policy is defined inside an element of `matchLabels` array, I would expect it only to configure the behavior for that single element. Take the following example:

```yaml
          matchLabels:
            - type: Value
              key: cloud-provider
              value: aws
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: spec.someValue
              fromFieldPathPolicy: Optional
```

I would expect this still to filter `cloud-provider=aws` if `spec.someValue` is absent.

> if a single label selector is specified and marked as optional, we should return no environment at all

This definitely makes sense, instead of selecting all by considering this as "no selector provided".

However, I see the inconsistency between the two cases, if a single element - we filter all out, if multiple elements, we filter all in from the entry with policy POV. 
Also it is not only 1 vs n selectors being provided, e.g. what happens if multiple - all optional / none resolved:

```yaml
          matchLabels:
            - type: FromCompositeFieldPath
              key: cloud-provider
              valueFromFieldPath: spec.anotherValue
              fromFieldPathPolicy: Optional
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: spec.someValue
              fromFieldPathPolicy: Optional
```

So, I agree that there is no perfect answer, but I still lean on policy impacting that single entry instead of invalidating all others. Otherwise, we may consider moving the policy up instead to make it more explicit:

```yaml
    environmentConfigs:
      - ref:
          name: example-environment-1
      - type: Selector
        selector:
          compositeValueResolutionPolicy: Optional # This environment config selector will be ignored if composite value resolution fails for any label
          matchLabels:
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: metadata.labels[stage]
```
